---
title: Downloading Climate Data from ERA5-Land
short_title: ERA5 Climate Data
---

In this notebook we will demonstrate how we can use Climate Tools and `dhis2eo` to retrieve climate data for a set of DHIS2 organisation units, based on the [ERA5-Land dataset](https://cds.climate.copernicus.eu/datasets/reanalysis-era5-land), hosted at the [ECMWF Climate Data Store (CDS)](https://cds.climate.copernicus.eu/datasets). 

----------------------------------------
## Requirements 

### 1. Create an ECMWF User

Before you can download the dataset programmatically, you need to [create an ECMWF user](https://www.ecmwf.int/user/login).

### 2. Authenticate with your ECMWF user

Next, you need to authenticate with this user based on your user credentials:

- Go to the [CDSAPI Setup page](https://cds.climate.copernicus.eu/how-to-api) and make sure to login.
- Once logged in, scroll down to the section "Setup the CDS API personal access token". 
  - This should show your login credentials, and look something like this:

        url: https://cds.climate.copernicus.eu/api
        key: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

- Copy those two lines to a file `.cdsapirc` in your user's $HOME directory.

### 3. Accept the dataset license

Lastly, ECMWF requires that you manually accept the user license for each dataset that you download. 

- Visit the Download page of the dataset you want to download (links below).
- Scroll down until you get to the "Terms of Use" section.
- Click the button to accept and login with your user if you haven't already. 

--------------------------------------------
## What you need

We start by importing the necessary libraries:

In [1]:
import dhis2eo
import dhis2eo.org_units
from dhis2eo.data.cds import era5_land

We also load our Sierra Leone organisation unit GeoJSON file downloaded from DHIS2. Alternative, see our [guide for how to fetch organisation units directly from your DHIS2 instance](../../org-units/download-web-api.ipynb). 

In [2]:
org_units = dhis2eo.org_units.from_file('../../data/sierra-leone-districts.geojson', dhis2=True)
org_units

org_unit_id          name  level  \
0   O6uvpzGd5pu            Bo      2   
1   fdc6uOvgoji       Bombali      2   
2   lc3eMKXaEfw        Bonthe      2   
3   jUb8gELQApl      Kailahun      2   
4   PMa2VCrupOd        Kambia      2   
5   kJq2mPyFEHo        Kenema      2   
6   qhqAxPSTUXp     Koinadugu      2   
7   Vth0fbpFcsO          Kono      2   
8   jmIPBj66vD6       Moyamba      2   
9   TEQlaapDQoK     Port Loko      2   
10  bL4ooGhyHRQ       Pujehun      2   
11  eIQbndfxQMb     Tonkolili      2   
12  at6UHUQatSo  Western Area      2   

                                             geometry  
0   POLYGON ((-11.5914 8.4875, -11.5906 8.4769, -1...  
1   POLYGON ((-11.8091 9.2032, -11.8102 9.1944, -1...  
2   MULTIPOLYGON (((-12.5568 7.3832, -12.5574 7.38...  
3   POLYGON ((-10.7972 7.5866, -10.8002 7.5878, -1...  
4   MULTIPOLYGON (((-13.1349 8.8471, -13.1343 8.84...  
5   POLYGON ((-11.3596 8.5317, -11.3513 8.5234, -1...  
6   POLYGON ((-10.585 9.0434, -10.5877 9.0432, -10...  
7   POLYGON ((-10.585 9.0434, -10.5848 9.0432, -10...  
8   MULTIPOLYGON (((-12.6351 7.6613, -12.6346 7.66...  
9   MULTIPOLYGON (((-13.119 8.4718, -13.1174 8.470...  
10  MULTIPOLYGON (((-11.5346 6.944, -11.5338 6.943...  
11  POLYGON ((-11.3546 8.6455, -11.357 8.6345, -11...  
12  MULTIPOLYGON (((-13.2435 8.1007, -13.2429 8.10...

--------

## Downloading ERA5-Land data

The `dhis2eo.data.cds.era5_land` module allows you to easily download climate data for different time periods from the ERA5-Land dataset. 

### Hourly data

To import daily ERA5-Land data into DHIS2, we recommend downloading [ERA5-Land data at hourly levels](https://cds.climate.copernicus.eu/datasets/reanalysis-era5-land), and then aggregating to up to the daily level. This is because the [daily version of ERA5-Land](https://cds.climate.copernicus.eu/datasets/derived-era5-land-daily-statistics) does not include the precipitation variable. 

The data can be downloaded by calling the function `dhis2eo.data.cds.era5_land.get('hourly', ...)`. By default, this downloads the most commonly requested ERA5 climate variables: 2m temperature and total precipitation, but you can also set the `variables` parameters to any of the ERA5-Land variables. 

Due to the large amounts of data, this dataset only allows downloading one month at a time. We therefore need to loop over each year and month, and call the download function for each `year` and `month` value. 

For the purpose of this notebook, let's download data for the last 3 months: 

In [4]:
start_year = 2025
start_month = 7
end_year = 2025
end_month = 9

for year in range(start_year, end_year+1):
    for month in range(1, 12+1):
        
        # skip months before or after our defined time range
        if (year,month) < (start_year,start_month):
            continue
        if (year,month) > (end_year,end_month):
            continue

        # download the climate data
        # commented out for this notebook
        print(f'Month: {year}-{month}')
        hourly_data = era5_land.get('hourly', year=year, month=month, org_units=org_units)

        # do something with the data
        # e.g. save to disk, aggregate, or import to DHIS2
        # ...

Month: 2025-7
dhis2eo.data.utils - INFO - Loading from cache: /tmp/dhis2eo_data_cds_era5_land_get_c4ea5010a0.nc
Month: 2025-8
dhis2eo.data.utils - INFO - Loading from cache: /tmp/dhis2eo_data_cds_era5_land_get_0dfabb33c4.nc


Month: 2025-9
dhis2eo.data.utils - INFO - Loading from cache: /tmp/dhis2eo_data_cds_era5_land_get_f079392f49.nc


To inspect the contents of the downloaded data, let's view the data of the last element in the loop: 

In [5]:
hourly_data

<xarray.Dataset> Size: 6MB
Dimensions:     (valid_time: 720, latitude: 31, longitude: 31)
Coordinates:
  * valid_time  (valid_time) datetime64[ns] 6kB 2025-09-01 ... 2025-09-30T23:...
  * latitude    (latitude) float64 248B 9.918 9.818 9.718 ... 7.118 7.018 6.917
  * longitude   (longitude) float64 248B -13.3 -13.2 -13.1 ... -10.5 -10.4 -10.3
    number      int64 8B ...
    expver      (valid_time) <U4 12kB ...
Data variables:
    t2m         (valid_time, latitude, longitude) float32 3MB ...
    tp          (valid_time, latitude, longitude) float32 3MB ...
Attributes:
    GRIB_centre:             ecmf
    GRIB_centreDescription:  European Centre for Medium-Range Weather Forecasts
    GRIB_subCentre:          0
    Conventions:             CF-1.7
    institution:             European Centre for Medium-Range Weather Forecasts
    history:                 2025-11-13T19:24 GRIB to CDM+CF via cfgrib-0.9.1...

We see that this is hourly data for the month of September 2025, and contains data variables `t2m` (2m temperature), and `tp` (total precipitation). For more information on how to aggregate from hourly to daily data, see our [guide to aggregate hourly to daily data](../../aggregation/time-period-aggregation.ipynb).  

### Monthly data

In some cases you may only be interested in monthly climate data at longer time scales. In those cases we recommend downloading the [monthly ERA5-Land data](https://cds.climate.copernicus.eu/datasets/reanalysis-era5-land-monthly-means). 

These data can be downloaded by calling the function `dhis2eo.data.cds.era_land.get('monthly', ...)`. Similarly to the hourly data, the default variables are temperature and precipitation, but this can be specified with the `variables` parameter. 

Since monthly data takes a lot less space, you can download the data by specifying a list of `years` and a list of `months`. 

Let's download monthly data for the last 10 years:

In [7]:
years = list(range(2015, 2025+1))
months = list(range(1, 12+1))
monthly_data = era5_land.get('monthly', years=years, months=months, org_units=org_units)

dhis2eo.data.utils - INFO - Loading from cache: /tmp/dhis2eo_data_cds_era5_land_get_efee0c4c8a.nc


And inspect the results:

In [8]:
monthly_data

<xarray.Dataset> Size: 1MB
Dimensions:     (valid_time: 130, latitude: 31, longitude: 31)
Coordinates:
  * valid_time  (valid_time) datetime64[ns] 1kB 2015-01-01 ... 2025-10-01
  * latitude    (latitude) float64 248B 9.918 9.818 9.718 ... 7.118 7.018 6.917
  * longitude   (longitude) float64 248B -13.3 -13.2 -13.1 ... -10.5 -10.4 -10.3
    number      int64 8B ...
    expver      (valid_time) <U4 2kB ...
Data variables:
    t2m         (valid_time, latitude, longitude) float32 500kB ...
    tp          (valid_time, latitude, longitude) float32 500kB ...
Attributes:
    GRIB_centre:             ecmf
    GRIB_centreDescription:  European Centre for Medium-Range Weather Forecasts
    GRIB_subCentre:          0
    Conventions:             CF-1.7
    institution:             European Centre for Medium-Range Weather Forecasts
    history:                 2025-11-13T19:00 GRIB to CDM+CF via cfgrib-0.9.1...

We see the result contains 130 months (over 10 years) of data for temperature and precipitation variables. 

## Next steps

This notebook has showed how to download the ERA5 climate data. For guidance on what to do with the data after downloading, see: 

* [Aggregate the data to your organisation units](../../aggregation/org-unit-aggregation.ipynb).
* [Import the aggregated data to DHIS2](../../import-data/using-python-client.ipynb).

As these are daily updated data, this script should be run regularly to continue feeding DHIS2 with the latest data. This can be done by automatically triggering the script to be run at regular intervals, e.g. via a `cron` job (guide to be added in the future). 